RetainOnDelete docs (#1059)
pulumi/pulumi-hugo: themes/default/content/docs/intro/concepts/resources/options/retainOnDelete.md

---
title: "RetainOnDelete"

> meta_desc: The retain resource option prevents calling the providers Delete function.

Reviewer: ```suggestion
meta_desc: The `retainOnDelete` resource option marks a resource to be retained during a delete operation.
```

menu:
  intro:
    identifier: retainOnDelete
    parent: options
    weight: 13
---

The `retainOnDelete` resource option marks a resource to be retained. If a resource is retained then when

> Pulumi deletes or replaces the resource it will not call through to the resource providers `Delete` method.

Reviewer: Perhaps add - "As a result, the resource will not be deleted from the backing cloud provider, but will be removed from the Pulumi state".

> Pulumi deletes or replaces the resource it will not call through to the resource providers `Delete` method.

Reviewer: ```suggestion
Pulumi deletes or replaces the resource it will not call through to the resource provider's `Delete` method.
```

This option can be used for shared resources that other stacks might be making use of.

To actually delete a retained resource, it must first be *unretained*.

* Set `retainOnDelete: false` and then run `pulumi up`

Once the resource is unretained, it can be fully deleted as part of a following update.

> If a retained resource is deleted by Pulumi and you want to actually delete it you will either need to use

Reviewer: ```suggestion
If a retained resource is deleted by Pulumi and you later want to actually delete it from the backing cloud provider you will either need to use
```

> your providers manual interface to find and delete the resource, or import the resource back into Pulumi then

Reviewer: ```suggestion
your provider's manual interface to find and delete the resource, or import the resource back into Pulumi then
```

---
title: "RetainOnDelete"
meta_desc: The `retainOnDelete` resource option marks a resource to be retained during a delete operation.
menu:
  intro:
    identifier: retainOnDelete
    parent: options
    weight: 13
---

The `retainOnDelete` resource option marks a resource to be retained. If a resource is retained then when

> Pulumi deletes or replaces the resource it will not call through to the resource provider's `Delete` method.

Reviewer 2: Nit: I think we can make this sentence a bit more clear. Something like "If this option is set, then Pulumi will not call through to the resource provider's Delete method during `pulumi destroy` operation.".

As a result, the resource will not be deleted from the backing cloud provider, but will be removed from the
Pulumi state.

> This option can be used for shared resources that other stacks might be making use of.

Reviewer 2: Is this an example of when people want to do that? This sentence feels a bit hanging out of context and may be hard to grasp.

To actually delete a retained resource, it must first be *unretained*.

> * Set `retainOnDelete: false` and then run `pulumi up`

Reviewer 2: How does that work? We've already destroyed the stack, so the resource is not in the state anymore. How can we unretain it? Will we import if from the cloud? Otherwise, I'm not sure what "unretained" means.

Author: Yeh maybe this should be flipped ordering with the paragraph below.

Once the resource is unretained, it can be fully deleted as part of a following update.

If a retained resource is deleted by Pulumi and you later want to actually delete it from the backing cloud provider you will either need to use
your provider's manual interface to find and delete the resource, or import the resource back into Pulumi then
unretain and delete it.

The default is to inherit this value from the parent resource, and `false` for resources without a parent.

{{< chooser language "javascript,typescript,python,go,csharp" >}}

{{% choosable language javascript %}}

```javascript

> let db = new Database("db", {}, { retainOnDelete: true});

Reviewer 2: ```suggestion
let db = new Database("db", {}, { retainOnDelete: true });
```

```

{{% /choosable %}}
{{% choosable language typescript %}}

```typescript

> let db = new Database("db", {}, { retainOnDelete: true});

Reviewer 2: ```suggestion
let db = new Database("db", {}, { retainOnDelete: true });
```
